xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/xwiki.js

```diff
@@ -1992,3 +1992,45 @@ document.observe("xwiki:dom:loaded", function() {
   interceptXMLHttpRequest();
   interceptFetch();
 })();
+
+/**
+ * Initializes the submit-once protection on the current page.
+ * Search for all form elements with a .xwiki-submit-once class one the document (or on an updated sub-part of the dom).
+ * For each form, once the form is submitted, mark it with the .xwiki-form-submitting class and prevent new submissions.
+ *
+ *  @since 14.10.17
+ *  @since 15.5.3
+ *  @since 15.8RC1
+ */
+require(['jquery'], ($) => {
+  /**
+   * Initializes the root element by adding event listeners to all forms with the class 'form.xwiki-submit-once'.
+   * When a form is submitted, it prevents the submission if it is already being submitted and adds a class to visually
+   * convey that the form was submitted.
+   *
+   * @param {Element} rootElement - The root element where the forms are located.
+   */
+  function init(rootElement) {
+    Array.from(rootElement.getElementsByClassName('xwiki-submit-once')).forEach(form => {
+      const submittingClass = 'xwiki-form-submitting';
+      form.classList.remove(submittingClass);
+      // Use a submit event handler with namespace 'xwiki-submit-once'. Before defining a handler, any potentially 
+      // registered listener is removed to prevent double registration of the same handler.
+      $(form).off("submit.xwiki-submit-once").on("submit.xwiki-submit-once", (e) => {
+        // Prevent if already submitting.
+        if (form.classList.contains(submittingClass)) {
+          e.preventDefault();
+        } else {
+          // Add a class to visually convey that the form was submitted
+          form.classList.add(submittingClass);
+        }
+      })
+    });
+  }
+
+  $(document).on('xwiki:dom:loaded xwiki:dom:updated', (event, data) => {
```

Comment:
I don't think you need to listen to ``xwiki:dom:loaded`` if you call the ``init`` function below using ``$(...)``. Basically you need to initialize the forms available on page load and those added dynamically afterward, for which a ``xwiki:dom:updated`` is triggered.